### PR TITLE
Allow to omit type of two way bindings property

### PR DESCRIPTION
**Pull request by ogoffart**

### CODE REVIEW

File: sixtyfps_compiler/passes/infer_aliases_types.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            let node = syntax_nodes::TwoWayBinding::new(node.clone())
+                .expect("The parser only avoid missing types for two way bindings");
+            let mut lookup_ctx = LookupCtx::empty_context(type_register, diag);
+            lookup_ctx.property_name = Some(prop);
+            let mut scope = scope.0.clone();
+            scope.push(elem.clone());
+            lookup_ctx.component_scope = &scope;
+            Expression::from_two_way_binding(node, &mut lookup_ctx)
+        }
+        _ => panic!("The parser should make sure that the expression is "),
```
Comment (tronical) [inline sixtyfps_compiler/passes/infer_aliases_types.rs:84]:
```suggestion
        _ => panic!("The parser should make sure that the expression is uncompiled"),
```

File: sixtyfps_compiler/passes/resolving.rs [hunk 2]
```diff
@@ -64,6 +64,10 @@ fn resolve_expression(
                 Expression::from_binding_expression_node(node.clone(), &mut lookup_ctx)
             }
             SyntaxKind::TwoWayBinding => {
+                if lookup_ctx.property_type == Type::Invalid {
+                    assert!(diag.has_error());
```
Comment (tronical) [inline sixtyfps_compiler/passes/resolving.rs:68]: This assert could perhaps benefit from a comment, elaborating how untyped two-way bindings start with Type::Void, get processed by the alias type inference pass by changing to Type::Invalid and attempting resolution. Failure to resolution must produce an error.

File: docs/langref.md [hunk 3]
```diff
@@ -168,11 +168,14 @@ Example := Button {
 Using the `<=>` syntax, one can create two ways binding between properties. These properties are now linked
 together.
 The right hand side of the `<=>` must be a reference to a property of the same type.
+The type can be omitted in a property declaration to have the type automatically infered.
```
Comment (tronical) [inline docs/langref.md:171]:
```suggestion
The type can be omitted in a property declaration to have the type automatically inferred.
```

File: docs/langref.md [hunk 4]
```diff
@@ -168,11 +168,14 @@ Example := Button {
...
 
 
 ```60
 Example := Window {
     property<brush> rect_color <=> r.background;
+    // it is allowed to omit the type to have it automatically infered
```
Comment (tronical) [inline docs/langref.md:177]:
```suggestion
    // it is allowed to omit the type to have it automatically inferred
```

File: sixtyfps_compiler/langtype.rs [hunk 5]
```diff
@@ -221,6 +221,7 @@ impl Type {
                 | Self::Struct { .. }
                 | Self::Array(_)
                 | Self::Brush
+                | Self::Void // Void is not a valid type for property, but it is temporarily used for infered two ways bindings
```
Comment (tronical) [inline sixtyfps_compiler/langtype.rs:224]:
```suggestion
                | Self::Void // Void is not a valid type for property, but it is temporarily used for inferred two ways bindings
```

**Review by tronical — APPROVED** [review]

File: sixtyfps_compiler/passes/resolving.rs (hunk 2 above)
Comment (ogoffart) [reply sixtyfps_compiler/passes/resolving.rs:68]:
```suggestion
                    // An attempt to resolve this already failed when trying to resolve the property type
                    assert!(diag.has_error());
```

File: sixtyfps_compiler/passes/infer_aliases_types.rs (hunk 1 above)
Comment (ogoffart) [reply sixtyfps_compiler/passes/infer_aliases_types.rs:84]:
```suggestion
        _ => panic!("There should be a Uncompiled expression at this point."),
```

**Pull request merged**